### PR TITLE
Definitions for inlining

### PR DESCRIPTION
**Pull request by springmeyer**
First attempt at tackling #28. There is such many resources on this. My goal here is to make the definitions:

 - very accessible to new developers who may be designing simple header-only libraries
 - a jumping off point to external resources that contain the gritty details
 - well interlinked with other glossary items (like header-only library)

/cc @GretaCB @daniel-j-h

### CODE REVIEW

**Comment by daniel-j-h** [comment]
Great and accessible docs! Not sure if we should split the description into
- the mechanical concept of inlining as a compiler optimization and
- linkage and when you have to use `inline` even if you don't care for performance

For the performance part maybe we could add a paragraph about how the compiler can not see across translation unit boundaries (it compiles a single cpp file plus all its includes separately - except when using link-time optimization). Putting function definitions into headers and making them inline let's the optimizer "see more" of your code and optimize more heavily based on that. Clang's `-Rpass` flag is great for debugging the inliner and what it does internally.

Related to that: once you start putting definitions into a header file (either because of performance reasons or e.g. because of templates) you need to mark them inline (technically not for non-fully specialized templates, but always inlining definitions in headers is a good default). Otherwise multiple object files will have the same definitions compiled into them. And this is undefined behavior (One-Definition-Rule - ODR - violation).

- https://clang.llvm.org/docs/UsersManual.html
- http://en.cppreference.com/w/cpp/language/storage_duration#Linkage

**Comment by springmeyer** [comment]
@daniel-j-h great great suggestions. Would you be interested in making those changes? If so, please do so directly to this PR and I'll review before merging. If not let me know and I'll try to integrate. At this point I'm less knowledge on ODR, so it would take me some time reading up on it to know how best to integrate.

**Review by daniel-j-h — COMMENTED** [review]
I just added more details, feel free to re-arrange and modify them.

File: glossary.md [hunk 1]
```diff
@@ -814,6 +814,129 @@ A custom allocator may be able to allocate memory faster, or use memory more eff
...
+}
+```
+
+Because in our example it knows that `result` will always be `1`.
+
+### inlined
+
+When a compiler [inline-expands](#inline-expands) a function.
+
+### definition
```
Comment (daniel-j-h) [inline glossary.md]: Ah just came across this while writing the odr section: we should split `definition` and `declaration`.

    int get();

is a declaration, while

    int get() { return 1; }

is a declaration.

http://en.cppreference.com/w/cpp/language/declarations
http://en.cppreference.com/w/cpp/language/definition

Comment (springmeyer) [reply glossary.md]: 👍  thanks. 79560afc9cc2fcb13 looks great, please merge once you've had a chance to fix this. Thanks!

Comment (daniel-j-h) [reply glossary.md]: rebased, added declaration vs definition, merged

Comment (springmeyer) [reply glossary.md]: 🎉  Thanks @daniel-j-h

**Pull request merged**